File: Java/src/main/java/QuestionBank/Solution1001_1100/Solution1030.java
package QuestionBank.Solution1001_1100;

import java.util.Arrays;
import java.util.Comparator;

/**
 * @author windows
 */
public class Solution1030 {
    public int[][] allCellsDistOrder(int rows, int cols, int rCenter, int cCenter) {
        int[][] res = new int[rows * cols][2];
        int index = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
               res[index++] = new int[]{i, j};
            }
        }
        Arrays.sort(res, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return Math.abs((o1[0] - rCenter) * (o1[0] - rCenter) + (o1[1] - cCenter) * (o1[1] - cCenter)) -
                        Math.abs((o2[0] - rCenter) * (o2[0] - rCenter) + (o2[1] - cCenter) * (o2[1] - cCenter));
            }
        });
        return res;
    }
    
    
    public static void main(String[] args) {
        Solution1030 solution1030 = new Solution1030();
        System.out.println(solution1030.allCellsDistOrder(1, 2, 0, 0));
    }
}

File: C++/src/Solution1001_1100/Solution1030.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution1030.h"
#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;


vector<vector<int>> allCellsDistOrder(int rows, int cols, int rCenter, int cCenter) {
  vector<vector<int>> res;
  queue<vector<int>> que;
  vector<vector<bool>> visited(rows,vector<bool>(cols, false));
  vector<vector<int>> dirs{
      {0, 1},
      {0, -1},
      {1, 0},
      {-1, 0}
  };
  que.push({rCenter, cCenter});
  visited[rCenter][cCenter] = true;
  while (!que.empty()){
    vector<int> nums = que.front();
    que.pop();
    res.push_back(nums);
    for (int i = 0; i < dirs.size(); ++i) {
      int a = nums[0] + dirs[i][0];
      int b = nums[1] + dirs[i][1];
      if (a >=0 && b >= 0 && a < rows && b < cols && !visited[a][b]){
        que.push({a, b});
        visited[a][b] = true;
      }
    }
  }
  return res;
}